src/main/java/software/amazon/smithy/lsp/language/RenameHandler.java

```diff
@@ -0,0 +1,260 @@
+        var config = ReferencesHandler.Config.create(project, idlFile, params.getPosition());
+        return getRenameRange(config.id().range(), config.id().copyIdValue());
+    }
+
+    /**
+     * @param params The request params
+     * @return A workspace edit that applies the rename
+     */
+    public WorkspaceEdit handle(RenameParams params) {
+        var config = ReferencesHandler.Config.create(project, idlFile, params.getPosition());
+        var edits = getEdits(config, params.getNewName());
+        return new WorkspaceEdit(edits);
+    }
+
+    private Map<String, List<TextEdit>> getEdits(ReferencesHandler.Config config, String newName) {
+        String namespace = config.shape().getId().getNamespace();
+
+        ShapeId renamedId;
+        try {
+            renamedId = ShapeId.fromRelative(namespace, newName);
+        } catch (ShapeIdSyntaxException e) {
+            throw invalidShapeId(e);
+        }
+
+        var projectEdits = new ProjectEdits(config, newName, renamedId, new HashMap<>());
+        projectEdits.collect(project);
+
+        return projectEdits.edits;
+    }
+
+    private record ProjectEdits(
+            ReferencesHandler.Config config,
+            String newName,
+            ShapeId renamedShapeId,
+            Map<String, List<TextEdit>> edits
+    ) {
+        private enum FileEditType {
+            CONFLICT,
+            SIMPLE
+        }
+
+        private void collect(Project project) {
+            var references = References.findReferences(config.model(), config.shape(), project);
+
+            addEdits(references);
+            deconflictDefinition();
+        }
+
+        private void addEdits(References allReferences) {
+            for (var fileReferences : allReferences.fileReferences()) {
+                FileEditType fileEditType = getEditType(fileReferences.idlFile());
+                if (fileEditType == FileEditType.CONFLICT) {
+                    addConflictRenames(fileReferences, renamedShapeId.toString());
+                } else {
+                    addSimpleRenames(fileReferences);
+                }
+            }
+
+            for (var definitionReference : allReferences.definitionReferences()) {
+                var uri = checkIfJar(definitionReference.idlFile());
+
+                addSimpleRename(uri, definitionReference.idlFile(), definitionReference.ref());
+            }
+        }
+
+        private void deconflictDefinition() {
+            var sourceFile = config.definitionFile();
+            if (sourceFile == null) {
+                return;
+            }
+
+            String conflictingId = getConflictingImport(sourceFile, newName);
+            if (conflictingId == null) {
+                return;
+            }
+
+            var conflictingShape = ShapeSearch.findShape(sourceFile.getParse(), conflictingId, config.model())
+                    .orElse(null);
+            if (conflictingShape == null) {
+                return;
+            }
```

Comment:
can this be improved? it looks awkward lol but maybe `Optional` is not that flexible